lxd_container module fails if certificate already in trust store #2750

june07 opened the pull request:
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

<!--- Name of the plugin/module/task -->

module_utils/lxd.py
affected lxd_container module
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.2.0

```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

When the client certificate is already stored, lxd returns a JSON error with message "Certificate already in trust store".  This "error" will occur on every task run after the initial run.  The cert should be in the trust store after the first run and this error message should really only be viewed as informational as it does not indicate a real problem.  I certainly don't think an exception should be being thrown?!

Fixes:
https://github.com/ansible/ansible-modules-extras/issues/2750

<!-- Paste verbatim command output below, e.g. before and after your change -->

```

```

udesilva committed:
Update playbooks_lookups.rst

Added a warning regarding attempting to create a local password file when the user the playbook is running as cannot create it... causes Ansible to loop at that task.

udesilva committed:
Update playbooks_lookups.rst

jinesh-choksi committed:
Key names in the ansible_local variable are converted into lower case

Adding a note to the documentation to highlight that key names in the ansible_local variable are converted into lower case when they are read from the local facts on the host.

Commit:
Updated grammar starting on line 166.

Commit:
lxd_container module fails if certificate already in trust store #2750

Commit:
no message

hnakamur commented on lib/ansible/module_utils/lxd.py:
```diff
@@ -121,6 +121,8 @@ def _send_request(self, method, url, body_json=None, ok_error_codes=None, timeou
             if resp_type == 'error':
                 if ok_error_codes is not None and resp_json['error_code'] in ok_error_codes:
                     return resp_json
+                if resp_json['error'] == "Certificate already in trust store":
```
@june07 Thanks for your pull request.

At first, I thought I do not like depending on the particular error message and it would be better to call the `GET /1.0/certificates` or `GET /1.0/certificates/<fingerprint>` beforehand to check the certificate is in trust store.

However, since the LXD REST API does not have concept of transaction, after checking with one of the two APIs and before calling `POST /1.0/certificates`, there may be other requests to register the same certificate.
So I think checking the error of `POST /1.0/certificates` is the way to go.

I looked at the LXD code and confirmed the error message is the only way to check the error type.
https://github.com/lxc/lxd/blob/26858d333f13ac5e97bbc40a542374c7ec9156c3/lxd/certificates.go#L147

So after all, I agree to your pull request.

june07 replied:
Thanks.

Commit:
Merge branch 'devel' of https://github.com/ansible/ansible into devel

jimi-c committed:
Create a raw lookup for hostvars that does not template the data

When using hostvars to get extra connection-specific vars for connection
plugins, use this raw lookup to avoid prematurely templating all of the
hostvar data (triggering unnecessary lookups).

Fixes #17024

machacekondra committed:
Add oVirt support to `ansible/module_utils` module (#17364)

* Add oVirt utility module

This patch add oVirt utility module, which contains helper functions,
for oVirt modules and also shared documentation fragment for oVirt.

* Adjust to Python 2.4

* Fixups

* Add support for poll interval and fixes

june07 committed:
Changes allow use of remote LXD containers which require a <remote lxd host>:<remote container> syntax.
Modification to lib/ansible/inventory/__init__.py adds a regexp to find ":::" which identifies a remote
LXD container, which is then translated to a single ":" by the lxd connection plugin before passing off
to lxc.

jimi-c committed:
Allow for empty source in dwim_path_relative_stack

Fixes #17192

mattclay committed:
Switch tests from sudo to become. (#17694)

dharmabumstead committed:
Merge pull request #15820 from jinesh-choksi/patch-2

Key names in the ansible_local variable are converted into lower case

supertom committed:
test_gce_mig integration test playbook (#17249)

Includes:
* Tests to create, delete, resize, recreate and configure an Autoscaler
* Tests for parameters only.  Use TEST_FLAGS="--tags 'param-check'" to execute only these tests.

Requires:
* An instance template to be configured in your project.

mattclay committed:
Remove tests for deprecated su and sudo. (#17697)

abadger committed:
Remove _load_hosts() from Play initialization as it's no longer needed and it breaks using extra_vars defining a list for hosts (#17699)

Thanks to @jimi-c for the solution

Fixes #16583

Christoph-D committed:
Add a test for int/float parameter type checking (#16741)

A parameter of type int should accept int and string, but not float.
A parameter of type float should accept float, int, and string.

Also reset the arguments in another test so that it runs cleanly.  This
agrees with what all the other tests are doing.

abadger committed:
Update submodule refs

privateip committed:
fixes mode on module_utils/dellos9.py (#17682)

pgrange committed:
fixes issue #13981: unsafe_writes block appeared too late in the atom… (#17405)

* fixes issue #13981: unsafe_writes block appeared too late in the atomic_move
workflow. This led to errno.EBUSY to not be managed in the context of
issue #!#981

* Reduce changes to fix #13981

* Abstract the unsafe_writes fallback into a helper method.
Explicitly try/except os.rename part of the code and call this helper method.
If the code fails in shutil.copy2 or shutil.move this should not be related to issue #13981
since they write to b_tmp_dest_name.
(as suggested by @abadger)

* Check if unsafe_writes in the caller, not in _unsafe_writes.
That way the function call reads as "Do an unsafe write"
and not as "I think we should do an unsafe_write.

jctanner committed:
ini lookup: add 'type' to the list of known arguments (#17707)

Fixes #16556

jimi-c committed:
Add max_passes to the debug strategy override of _process_pending_results

Fixes #17520

supertom committed:
Integration Playbook for Google Cloud DNS module. (#17698)

mattclay committed:
Add keep on failure option to integration.sh. (#17711)

alikins committed:
add a defaults for integration.sh (#17709)

* add a default test image (ansible/ansible:centos7)
* add a default test target (all)

alikins committed:
Add a 'make integration' target (#17710)

Runs test/utils/shippable/integration.sh

alikins committed:
Add '--cover-erase' to 'make tests' cli (#17708)

Otherwise the coverage is cumulative over multiple runs and can be misleading.

mattclay committed:
Add shellcheck to code-smell checks. (#17715)

privateip committed:
fixes issue where junos shared module was ignoring ssh_keyfile (#17712)

This fixes a problem with the Netconf transport in which the ssh keyfile
wasn't being used if it was defined.  The ref issue is filed against 2.1.1
but have been unable to replicate the problem in that version

ref: ansible/ansible-modules-core#4966

nitzmahone committed:
bump extras submodule ref

mattclay committed:
Update integration.sh to copy source by default. (#17717)

Also add more documentation to the script.

mattclay committed:
Remove run_tests.sh and update docs. (#17719)

Removed the obsolete test/utils/run_tests.sh script in favor of
using the newer test/utils/shippable/integration.sh instead.

bcoca committed:
clarified `omit` intended usage

fixes #13986

j0hnsmith committed:
Feature/add ansible play hosts all (#17498)

* refactor ignore_limits_and_restrictions

into ignore_limits and ignore_limitations

* add ansible_play_hosts_all

* update docs re ansible_play_hosts_all

* only use play.hosts when is has a value

* replace ansible_play_hosts with ansible_play_hosts_all

* remove unnecessary var

bcoca committed:
clarified play_hosts/ansible_play_hosts docs

bcoca committed:
added ansible_play_hosts to changelog

privateip committed:
fixes issue introduced with raw kwarg (#17728)

The raw kwarg was added to return raw output from devices with if the
attempt to convert to json failed.  The change was causing all json
output to be returned raw.  This fixes that issue.

Commit:
add hint to use hash filters for password in faq

bcoca committed:
added note about loops being lookups

june07 committed:
Base working dynamic inventory script and ini file.

dfarrell07 committed:
Remove deprecated roll path arg from docs #13563 (#17689)

Signed-off-by: Daniel Farrell <dfarrell@redhat.com>

dharmabumstead committed:
Merge pull request #17730 from davidak/hash

add hint to use hash filters for password in FAQ

dharmabumstead committed:
Minor editing of PR#17730

abadger committed:
Add tests for the mount module (#17718)

* Add tests for the mount module

* Switch from unmounted to absent...

the code for mounting always modifies fstab so we need to always modify
fstab to cleanup as well.

* Fix comments and copyright

abadger committed:
Update submodule refs to pick up mount fixes

hyeonseongc committed:
fix for rsync protocol support (#16756)

dharmabumstead committed:
Dev guide reorg continues (#17732)

mattclay committed:
Set PRIVILEGED=true for non_destructive tests. (#17733)

This should allow test_mount tests to run on Shippable.

mattclay committed:
Improve error handling for ansible-core-ci. (#17734)

abadger committed:
Update submodule refs

abadger committed:
Disable mount tests for now.  Mount is buggy on too many platforms

mattclay committed:
Disable expect continue in ansible-core-ci. (#17737)

This will allow use of larger SSH keys.

mattclay committed:
Revert "Set PRIVILEGED=true for non_destructive tests. (#17733)" (#17738)

This reverts commit 1384270ccd50cc7660c9375708fda148639ab797.

skg-net committed:
Update OS6 sublevels (#17735)

mscherer committed:
Add a umask argument to run_command

In order to avoid problem due to race conditions, it is
required to run umask when generating some sensitive files,
such as a TLS key.

dharmabumstead committed:
Fixed dev info landing page link

bcoca committed:
removed unused code, added comment explaining

files is really a placeholder for common code for separate service modules, was copy of current service module and this seemed to confuse people so this update should clear that up

cinerama70 committed:
gen_distribution_version_testcase.py should fail if ansible run fails (#17693)

Currently, "ansible localhost -m setup" can fail silently during the
run of gen_distribution_version_testcase.py, resulting in incorrect
output. Use check_output() rather than communicate() and handle
the exception if we get a nonzero return value.

privateip committed:
fixes issue where arguments for regexp search() where transposed (#17764)

The arguments for the regex search() function were transposed in the
netcli match() method that caused conditionals to fail. Switched the
arguments to fixe the bug

fixes #17749

abadger committed:
Update submodule refs (especially to pick up the mount fixes)

abadger committed:
Update submodule refs

bcoca committed:
fix delegate_facts (#17736)

code was misplaced when we changed result processing
also cleaned up import and clarified item/label
fixes #17582

bcoca committed:
undefined var should fail 'label' setting (#17675)

bcoca committed:
removed deprecated 'bare' templating environment (#17640)

https://docs.ansible.com/ansible/porting_guide_2.0.html#deprecated

bcoca committed:
removed deprecated first available file from tasks (#17643)

https://docs.ansible.com/ansible/porting_guide_2.0.html#deprecated

bcoca committed:
remove bare vars and undefined from lookup lists (#17647)

https://docs.ansible.com/ansible/porting_guide_2.0.html#deprecated

nitzmahone committed:
swallow errors during async polling (#17760)

also use connection reset on exceptions if available (eg, prevent reuse of "stuck" WinRM connections due to reboot/NIC bounce/etc)

mattclay committed:
Update submodule refs.

jimi-c committed:
Clear the plugin path cache when adding new directories

mattclay committed:
Revert "Clear the plugin path cache when adding new directories" (#17785)

This reverts commit 5a57313dd72e574d38272364df3ef01d6a3646ef.

mattclay committed:
Update get_url test to use httptester. (#17787)

privateip committed:
raises ValueError exception if conditional is not parsable (#17788)

The Conditional class now raises a ValueError with message if it cannot
correclty parse the passed in conditional.  This makes it easier to
detect issues in modules that specify conditionals.

alikins committed:
Add mnt pnts with single quote to test (16855) (#17771)

This adds some test data to test_facts.py that
includes mnt points that have a single quote in
the path.

Ala, https://github.com/ansible/ansible/issues/16855

The bug was already fixed via other changes, but this is
for regression testing.

mattclay committed:
Enable unarchive tests for OS X and FreeBSD. (#17789)

mattclay committed:
Run test_template tests on OS X. (#17791)

gundalow committed:
Notice of Networking *_template deprecations (#17668)

* Notice of Networking *_facts deprecations

* Just use the term "Deprecations"

* Update CHANGELOG.md

* Improve layout

abadger committed:
Use the play term instead of playbook.

An ansible playbook is a list of plays.  environment: applies per-play,
not per-playbook.

bcoca committed:
ansible_play_batch (#17779)

* ansible_play_batch

* added version added

bcoca committed:
better handling of host lists (#17781)

* better handling of host lists

fixes #17762

* corrected message

* fixed extras space

abadger committed:
The mode that we fake from stat is checked more closely under python3.5 on El Capitan (#17794)

Python2 seems to allow any integer.  Python3.5 on Linux seems to allow
a 32 bit unsigned int.  Python3.5 on El Capitan seems to limit it to
a smaller size...  perhaps a 16 bit int.

abadger committed:
Update submodules refs

abadger committed:
Turn mount test back on (#17797)

* Turn mount test back on

* Mount tests need PRIVILEGED so turn that back on

Revert "Revert "Set PRIVILEGED=true for non_destructive tests. (#17733)" (#17738)"

This reverts commit dc0fb1c212df5f08490c3e959b9dce64ff540169.

* Add a needs_privileged tag so that we can skip mount tests on centos6

Some containers timeout on shippable tests when run with privileged.
Unfortunately, some tests require that in order to run.  Tagging those
allows us to skip those tests on the platforms that timeout when we get
ready to run the integration test in shippable.

* Centos6 times out with PRIVILEGED set so remove that (will disable the mount tests on centos6)

* Remove false start

privateip committed:
adds overwrite kwarg to load_config in junos (#17798)

The junos load_config() method supports operations of overwrite, replace
and merge.  This adds the missing overwrite keyword arg to load_config()
so that action in junos_template can be procesed correctly.

dharmabumstead committed:
Fixes issue #17795

fajpunk committed:
Document `copy` module vault decryption behavior (#17777)

gillesgagniard committed:
fixes #16865 : we should properly determine connection type for delegated task first, before overriding remote user for local connections (#17723)

bcoca committed:
fixed typo

dharmabumstead committed:
Fixes issue #17759

cinerama70 committed:
Graceful error handling when delegate_to host null (#17453) (#17783)

Currently, if the host specified in delegate_to for a task is null,
Ansible will crash with a stack trace. Add a check for this state
and handle the error appropriately.

privateip committed:
fixes conditional processing with junos and xml data structures (#17801)

The conditional processing was failing due for two reasons:

1) The xml to json conversion string was not happening before the runner
   was processing the results
2) The Conditional instance was not parsing conditionals encoded with []

This fix address both issues.

abadger committed:
Update submodule refs

abadger committed:
Add PRIVILEGED to freebsd because testing of mount keys off of that. (#17803)

pkyeck committed:
Fix length of quickstart video mentioned in docs

mckerrj committed:
Update ROADMAP_2_2.rst

mckerrj committed:
Update ROADMAP_2_2.rst

alikins committed:
mv test_git to 'destructive' integration tests (#17526)

The test_git role removes ~/.ssh/known_hosts currently
and that is destructive.

Commit:
Use new YAML syntax in role dependencies.

bcoca committed:
add ssh error message to failure

jctanner committed:
Smart mode for sftp+scp (#17813)

If the sftp fails, roll over to scp by default. This saves users
from having to know about the scp_if_ssh method when sftp is broken
on the remote host.

jctanner committed:
Update changelog for scp_when_ssh=smart (#17819)

dharmabumstead committed:
Merge pull request #17809 from pkyeck/devel

Fix length of quickstart video mentioned in docs

dharmabumstead committed:
Merge pull request #14691 from udesilva/patch-1

Update Lookups documentation regarding a possible pitfall with password plugin

dharmabumstead committed:
Merge pull request #17556 from gahan-corporation/devel

Updated grammar starting on line 166.

jctanner committed:
vmware_inventory: Make the ceritifate check default, and create a config option (#17830)

Fixes #17811

resmo committed:
docsite: various fixes in loops doc

- "new in version ..." did not always show up
- indentation in examples
- minor style cleanup (whitespace)

Commit:
Merge pull request #17818 from chouseknecht/devel

Parse role dependencies consistently

gadost committed:
jabber callback (#16981)

* jabber callback

* delete empty lines

* grammar fix

* fix: do not print error when jabber callback disabled

* change callback version 1 -> 2

trondhindenes committed:
Bugfix: fixes new-alias bug when debugging powershell (#17816)

nitzmahone committed:
remove Windows async doc exclusion

dharmabumstead committed:
Merge pull request #17833 from resmo/docs/fix_loops

docsite: various fixes in loops doc

bcoca committed:
fix fact gathering flag

privateip committed:
fixes issue when calling save_config() and transport is nxapi (#17835)

An unhandled exeception is raised with using nxapi transport and setting
the save argument to true.  This fix will allow the configuration to be
saved regardless of the transport.

fixes ansible/ansible-modules-core#5094

bcoca committed:
dzdo prompt fix

fixed tests
dzdo has 2 paths now, with and w/o password
fixes #17838

privateip committed:
adds better error handling when conditional fails (#17839)

When the conditional cannot extract a value from the result string,
an unhandled exception would be raised.  This fix now gracefully handles
the exception

privateip committed:
fixes unicode conversation from junos get_config() method (#17841)

The junos config should convert the returning configuration to unicode
not str.  This fixes that issue.

jimi-c committed:
Fixing parentage of include_role objects for param inheritance

Fixes #17686

indrajitr committed:
Add homebrew in squash_actions list (#16966)

`homebrew`, like other package modules in the existing `squash_actions` list can
benefit from `with_items` loops optimization.

nitzmahone committed:
bump core submodule ref for async_wrapper fixes

mattclay committed:
Disable privileged mode on Shippable. (#17844)

skg-net committed:
Remove the dellosX_template module (#17836)

jimi-c committed:
Ensure the "attempts" key is set in the final result for do/until loops

Fixes #17258

nitzmahone committed:
bugfixes to JSON junk filter, added unit/integration tests to exercise (#17834)

nitzmahone committed:
bump core submodule ref for windows async_wrapper bugfix

privateip committed:
adds new exception for adding condition statements to Conditional (#17859)

When adding condition statements, the Conditional instance will now generate
an AddConditionError if is unable to map the condition to a function in the
instance

privateip committed:
switches the kickstart arg from True to False for eos (#17858)

The kickstart kwarg should be set to False for eos based devices and
was set to True.  This change cleans up problems loading json output
from cli commands

All eos_command test cases are now passing successfully

fixes #17441

privateip committed:
fixes issue with run_commands raising error (#17861)

The junos run_commands() method should raise an error when an RpcError is
returned but didn't when using display=text.  This fixes that error

abadger committed:
Normalize text and byte type in the ssh plugin helper method that builds up an ssh command (#17860)

Mostly cleanups to make the code more efficient, more pythonic, and obey
the unicode sandwich strategy more but also Fixes #17832

abadger committed:
On python3, subprocess needs another arg to pass extra file descriptors

agx committed:
Add foreman callback plugin (#17141)

abadger committed:
Update submodule refs

abadger committed:
Remove unicode escape (#17866)

* Remove unicode-escape which is not present on python3

Alternative fix for #17305

* Enable the assemble test on python3

* Fix other problems with assemble on python3

bcoca committed:
only change dir to playdir if local

fixes #17869
fixes #17770

bcoca committed:
added forman callback plugin to changelog

Commit:
added network updates to CHANGELOG

gundalow committed:
New modules (#17870)

Full review of new modules

nitzmahone committed:
bump core submodule ref

jimi-c committed:
Updating CHANGELOG and release versions for 2.1.2

jimi-c committed:
New release v2.2.0.0-0.1.rc1

Commit:
Merge branch 'devel' of github.com:june07/ansible into devel

alikins commented:
Not sure what happened to the branch, but it needs some attention to the git branching.

june07 commented:
Closing.  See https://github.com/ansible/ansible/pull/18115 instead.
The commit history got messed up so I had to resubmit the PR as above.
